File: solutions/_496_next_greater_element_i/solution.java
package _496_next_greater_element_i;

import java.util.*;

class Solution {

  public int[] nextGreaterElement(int[] a, int[] b) {
    HashMap<Integer, Integer> val2idx = new HashMap<>();
    for (int i = 0; i < a.length; i++) {
      val2idx.put(a[i], i);
    }

    int[] res = new int[a.length];
    Arrays.fill(res, -1);
    Stack<Integer> stack = new Stack<>();

    for (int i = 0; i < b.length; i++) {
      int cur = b[i];

      while (stack.size() > 0 && cur > stack.peek()) {
        int val = stack.pop();
        int idx = val2idx.get(val);
        res[idx] = cur;
      }

      if (val2idx.containsKey(cur)) {
        stack.push(cur);
      }
    }

    return res;
  }
}
/*
https://leetcode.com/problems/next-greater-element-i/
*/

File: solutions/_496_next_greater_element_i/solution.py
from typing import List


class Solution:
    def nextGreaterElement(self, a: List[int], b: List[int]) -> List[int]:
        val2idx = {n: i for i, n in enumerate(a)}
        res = [-1] * len(a)

        stack = []
        for i in range(len(b)):
            cur = b[i]

            while stack and cur > stack[-1]:
                val = stack.pop()
                idx = val2idx[val]
                res[idx] = cur

            if cur in val2idx:
                stack.append(cur)

        return res


# https://leetcode.com/problems/next-greater-element-i/
